scopegen/parser/parser.go

```diff
@@ -0,0 +1,181 @@
+	Name   string
+	Scopes []string
+}
+
+// File represents a proto file with all OAuth2 information parsed.
+// Scopes and methods are converted from map to sorted slice of struct
+// to have determinate generated codes.
+type File struct {
+	Package string
+	// Scopes is list of scopes defined for the service.
+	Scopes []Scope
+	// Methods is mapping between full method name and its scopes.
+	Methods []Method
+}
+
+func (f *File) validate() error {
+	for _, method := range f.Methods {
+		for _, scope := range method.Scopes {
+			var defined = false
+			for _, definedScope := range f.Scopes {
+				if definedScope.Name == scope {
+					defined = true
+					break
+				}
+			}
+			if !defined {
+				return fmt.Errorf("failed to parse, scope %q of method %q is undefined", scope, method.Name)
+			}
+		}
+	}
+
+	return nil
+}
+
+// generateFullMethodName generates the full method name of a gRPC service method.
+// It is reserve-engineered from protoc-gen-go.
+// Reference: https://github.com/golang/protobuf/blob/822fe56949f5d56c9e2f02367c657e0e9b4d27d1/protoc-gen-go/grpc/grpc.go
+func generateFullMethodName(pkg, servName, methName string) string {
+	var fullServName = servName
+	if pkg != "" {
+		fullServName = pkg + "." + fullServName
+	}
+	return fmt.Sprintf("/%s/%s", fullServName, methName)
+}
+
+func parseScopes(file *descriptor.FileDescriptorProto) (string, []Scope, error) {
+	var (
+		oauth2SecuritySchemeName string
+		scopes                   []Scope
+	)
+
+	fileExt, err := proto.GetExtension(file.GetOptions(), options.E_Openapiv2Swagger)
+	if err == proto.ErrMissingExtension {
+		// no swagger option defined, generated nothing
+		return "", nil, nil
+	} else if err != nil {
+		return "", nil, err
+	}
+
+	swagger := fileExt.(*options.Swagger)
+
+	if swagger.GetSecurityDefinitions() != nil {
+		for securitySchemeName, securityScheme := range swagger.GetSecurityDefinitions().GetSecurity() {
+			if securityScheme.Type == options.SecurityScheme_TYPE_OAUTH2 {
+				if securityScheme.Scopes != nil {
+					var scopeKeys []string
+					for scopeName := range securityScheme.Scopes.Scope {
+						scopeKeys = append(scopeKeys, scopeName)
+					}
+					sort.Strings(scopeKeys)
+
+					for _, scopeName := range scopeKeys {
+						scopes = append(scopes, Scope{
+							Name:  scopeName,
+							Value: securityScheme.Scopes.Scope[scopeName],
+						})
+					}
+
+					oauth2SecuritySchemeName = securitySchemeName
+					// for now we only handle first OAuth2 security scheme definition,
+					// supports for generating multiple schemes might come later
+					break
+				}
+			}
+		}
+	}
+	return oauth2SecuritySchemeName, scopes, nil
+}
+
+func parseMethods(securitySchemeName string, file *descriptor.FileDescriptorProto) ([]Method, error) {
+	var (
+		methods = make(map[string][]string)
+		result  []Method
+	)
+
+	if securitySchemeName == "" {
+		return nil, nil
+	}
+
+	for _, service := range file.GetService() {
+		for _, method := range service.GetMethod() {
```

Comment:
nit: this block is meaty enough to warrant its own function IMO (`extractMethodsFromService`)